Implement web client

zhammer opened the pull request:
Would love advice on how to make the react client as 'good react' as possible. 

Some things I may do:
- [x] Change the components to functional components
- [x] probably should add proptypes

Commit:
Create basic React App for finneganforever.com

Commit:
Update Passage component styling

Use 'Cormorant Garamond' font

Commit:
Add serverless-finch plugin for web_client deployment to s3.

Commit:
Add mechanical theater display inspired by VA opera exhibit

Commit:
Update icon and manifest.json metadata

Commit:
Add gif demo to README

Commit:
Animate clouds and waves in MechanicalTheatre

Commit:
Redirect all paths to finneganforever.com/#!

I tried to redirect all paths simply to finneganforever.com, but
replaceKeyWith can't be an empty string. '#!' comes from a suggestion
from
https://stackoverflow.com/questions/16267339/s3-static-website-hosting-route-all-paths-to-index-html.

evanhammer commented on web_client/src/Passage.jsx:
```diff
@@ -0,0 +1,33 @@
+import './Passage.css';
+
+class Passage extends Component {
+    componentWillMount() {
```
use `componentDidMount` for data fetching

evanhammer commented on web_client/src/Passage.jsx:
```diff
@@ -0,0 +1,33 @@
+import readPassage from './api';
+import './Passage.css';
+
+class Passage extends Component {
```
```js
state = {
  passageText: '',
}

evanhammer commented on web_client/src/Passage.jsx:
```diff
@@ -0,0 +1,33 @@
+class Passage extends Component {
+    componentWillMount() {
+        this.setState({passageText: ''});
+        readPassage(this.props.finneganForeverUrl)
```
`readPassage` should know about `finneganForeverUrl` so this component doesn't have to

evanhammer commented on web_client/src/Passage.jsx:
```diff
@@ -0,0 +1,33 @@
+            .then(({ passage, reading_every }) => {
+                this.setState({passageText: passage});
+                const readingEvery = reading_every * 1000;
+                setInterval(
```
look into memory leaks? https://heichwald.github.io/2016/01/10/memory-leak-closure-javascript.html

evanhammer commented on web_client/src/Passage.css:
```diff
@@ -0,0 +1,20 @@
+
+@media screen and (min-width: 960px) {
+  .text {
+    font-size: 48px;
```
60px? check out `rem`

zhammer commented on web_client/src/App.js:
```diff
@@ -0,0 +1,20 @@
+        return (
+            <div className="App">
+              <MechanicalTheatre>
+                <Passage finneganForeverUrl={FINNEGAN_FOREVER_URL}/>
```
A little confused re: our conversation earlier. Is the idea here to do:

```jsx
<IntervalTextReader readText=readText>
    <Passage /> // <--- the style component
</IntervalTextReader>
```

@evanhammer

evanhammer replied:
yeah, that's one approach. You'll have to add props to `Passage` within IntervalTextReader. The other is a higher order component: `const IntervalTextReader = intervalTextReader(Passage);`

https://medium.com/ingenious/a-tale-of-higher-order-components-render-props-a1ba47e8cfeb

zhammer replied:
Sweet that seems like it makes more sense

Commit:
Delete create-react-app README.md

Commit:
Refactor Passage to take a `readPassage` function for reading text

Convert App.js to a functional componenet

Commit:
Convert MechanicalTheatre to a functional component

Commit:
Update web client to use 'FW' logo

zhammer commented:
Still need to figure out what is the plan for `IntervalTextReader` / `Passage`.

Commit:
Remove unused { Component } imports

zhammer commented:
https://reactjs.org/docs/higher-order-components.html

`<PassageWithIntervalReader>`

Commit:
Add `withTextIntervalReader` HOC for fin passage fetching

Commit:
Add documentation to components

Add documentation to API functions

Delete package-lock.json as it is redundant when we have yarn.lock

Add documentation to App.js

Commit:
Switch componentDidUnmount to componentWillUnmount

zhammer commented on web_client/src/textIntervalReader.jsx:
```diff
@@ -0,0 +1,42 @@
+ *  of seconds after which new `text` should be read. Every `readingEvery` seconds, this component
+ *  will read text from `readTextInterval` and pass the read text to `WrappedComponent`s `text` prop.
+ */
+const withTextIntervalReader = (WrappedComponent, readTextInterval) => (
```
@evanhammer let me know if you have any feedback on this. three things i'd like feedback on

1. Naming (i.e. `onTextIntervalRead`) and signature (decomposing the promise's output rather than taking in specified parameters). Would you change either?
2. Not many of the helper functions here are really testable, since the main 'I/O' is setting the component state, and there's almost no business logic (besides setting up an interval `onFirstTextIntervalRead`.
3. I get this warning when running `yarn test`
```
  console.error node_modules/fbjs/lib/warning.js:33
    Warning: Can't call setState (or forceUpdate) on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
        in Unknown (at App.js:17)
        in div (at MechanicalTheatre.jsx:21)
        in div (at MechanicalTheatre.jsx:12)e-
        in MechanicalTheatre (at App.js:16)
        in div (at App.js:15)
        in App (at App.test.js:7)
```
A stackoverflow thread mainly concluded on: "this is why you keep async functions outside of the component", but I don't know how to do that while setting state. Any ideas?

zhammer replied:
Re: 3, it seems to me that I _am_ essentially cancelling all async operations on `componentWillUnmount` when I clear the interval. Maybe the linter isn't picking up on this?

evanhammer replied:
RE Naming. I like `withTextIntervalReader` - I like the `with` prefix for HOC's

evanhammer replied:
Oh, you were asking about the function names and signatures. not the HOC...

evanhammer replied:
RE Testing. I'd test to make sure that the interval fires regularly, and that state is updated. (or just that state is updated regularly)

evanhammer replied:
RE memory. I think what is happening is when the component unmounts, the intervals is cleared, but if the request to the server has already fired then when it returns you're calling setState on an unmounted component.

Options:
* Store the state outside of this component. Perhaps in App.
* Check to make sure the component is still mounted before calling setState.
* Look up some other option on the Internet.

zhammer replied:
re: memory, since it's a no op if setState is fired and a rare one at that, i think i'll leave it alone for now

evanhammer commented on web_client/src/textIntervalReader.jsx:
```diff
@@ -0,0 +1,42 @@
+        state = {
+            text: ''
+        }
+        onTextIntervalRead({ text }) {
```
with a single parameter, I'd prolly just pass it: `onTextIntervalRead(text)`

evanhammer replied:
I like `on` for event handlers. Though I also use a weird naming convention some time: past tense. It makes it clear that the event already happened.

zhammer replied:
This is past tense. 'read' as in, we have 'read' the text interval. I just like `on` since it makes it clear it's an event handler.

I do realize it's a little confusing because 1) 'read' is ambiguous and 2) `onButtonClick` is the standard, usually starts with on and is a present tense verb. In my ideal world it's `onButtonClicked`.

evanhammer commented on web_client/src/textIntervalReader.jsx:
```diff
@@ -0,0 +1,42 @@
+
+withTextIntervalReader.propTypes = {
+    WrappedComponent: PropTypes.element.isRequired,
+    readTextInterval: PropTypes.func.isRequired
```
I would name this function differently. This function doesn't care about the interval aspect. It's just `readText`. And it happens to be called at an interval.

So I'd also call the event handler `onTextRead`.

That said, I'd prolly do the past tense thing here: `readTextRequested` and `readTextSucceeded`.

zhammer replied:
What do you mean it "doesn't care about the interval aspect"? `readTextInterval` returns `text` and returns the `reading_every` interval.

zhammer replied:
That being said, it could be nice if the signature were

```jsx
const withTextIntervalReader = (WrappedComponent, readText, interval) => (...);

withTextIntervalReader.propTypes = {
    WrappedComponent: PropTypes.element.isRequired,
    readTextInterval: PropTypes.func.isRequired,
    interval: PropTypes.number.isRequired
};
```

but then `App` has to know how to read `readingEvery` / `interval` from the finneganForever api. I do think that the component is much more reusable if it doesn't depend on a function returning `readingEvery` on a text read.

For the time being I think `readTextWithInterval` is a better name.

evanhammer commented on web_client/src/textIntervalReader.jsx:
```diff
@@ -0,0 +1,42 @@
+        }
+        onFirstTextIntervalRead({ text, readingEvery }) {
+            this.intervalId = setInterval(() => {
+                readTextInterval().then(this.onTextIntervalRead.bind(this));
```
instead of `bind` you can define (line 15) the function as an arrow function:

```js
onTextIntervalTead = (text) => {
  this.setState({ text });
}

evanhammer commented on web_client/src/textIntervalReader.jsx:
```diff
@@ -0,0 +1,42 @@
+            this.onTextIntervalRead({text});
+        }
+        componentDidMount() {
+            readTextInterval()
```
I'd also prolly make a local function called `readText` that componentDidMount and setInterval call:

```js
readText () => {
  this.readTextInterval()
    .then(text => this.setState({ text }));
}

zhammer replied:
I think you didn't realize that `readTextInterval` returns `{ text, readingEvery }`

Commit:
Use arrow functions instead of binds

Commit:
Add fade between text changes in Passage

Commit:
Add Mechanical Theatre reference to README
